protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -0,0 +1,99 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.processor.base;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableBiMap;
+
+import org.apache.james.mailbox.MessageUid;
+
+public class UidMsnMapper {
+
+    public final static int FIRST_MSN = 1;
+
+    private final HashBiMap<Integer, MessageUid> msnToUid;
+
+    public UidMsnMapper() {
+        msnToUid = HashBiMap.create();
```

Comment:
Is the previous code *atomic*? Really?

Ok for error handling, it's not cool, and I missed it.

This commit is from @strokyl originally and did some pair programming with you on it. I can perfectly drop it as well, however, I considered it was a dreamed occasion for it to make it's way to **master**.

 - It brings new tests, which led to many improvements to this class.
 - We have better concern isolation
 - I like recycling code waiting for merges.